src/github.com/getlantern/detour/detour.go

```diff
@@ -43,325 +38,326 @@ import (
-		if !whitelisted(addr) {
-			detector := blockDetector.Load().(*Detector)
-			dc.setState(stateInitial)
-			// always try direct connection first
-			dc.conn, err = net.DialTimeout(network, addr, TimeoutToDetour)
-			if err == nil {
-				if !detector.DNSPoisoned(dc.conn) {
-					log.Tracef("Dial %s to %s succeeded", dc.stateDesc(), addr)
-					return dc, nil
+func Dialer(detourDialer dialFunc) func(network, addr string) (net.Conn, error) {
+	return func(network, addr string) (net.Conn, error) {
+		dc := &Conn{
+			addr:            addr,
+			writeBuffer:     new(bytes.Buffer),
+			conns:           make(chan conn, 2),
+			chDetourConn:    make(chan conn),
+			chRead:          make(chan ioResult),
+			chWrite:         make(chan ioResult),
+			chDialDetourNow: make(chan bool),
+		}
+		// use buffered channel as we may send twice to it but only receive once
+		chAnyConn := make(chan bool, 1)
+		ch := make(chan conn)
+		// dialing sequence
+		if whitelisted(addr) {
+			dialDetour(network, addr, detourDialer, ch)
+		} else {
+			go func() {
+				dialDirect(network, addr, ch)
+				dt := time.NewTimer(DelayBeforeDetour)
+				select {
+				case <-dt.C:
+				case <-dc.chDialDetourNow:
 				}
-				log.Debugf("Dial %s to %s, dns hijacked, try detour", dc.stateDesc(), addr)
-				if err := dc.conn.Close(); err != nil {
-					log.Debugf("Unable to close connection: %v", err)
+				if dc.anyDataReceived() {
+					ch <- nil
+					return
 				}
-			} else if detector.TamperingSuspected(err) {
-				log.Debugf("Dial %s to %s failed, try detour: %s", dc.stateDesc(), addr, err)
-			} else {
-				log.Debugf("Dial %s to %s failed: %s", dc.stateDesc(), addr, err)
-				return dc, err
-			}
-		}
-		// if whitelisted or dial directly failed, try detour
-		dc.setState(stateDetour)
-		dc.conn, err = dc.dialDetour(network, addr)
-		if err != nil {
-			log.Errorf("Dial %s failed: %s", dc.stateDesc(), err)
-			return nil, err
+				dialDetour(network, addr, detourDialer, ch)
+			}()
 		}
-		log.Tracef("Dial %s to %s succeeded", dc.stateDesc(), addr)
-		if !whitelisted(addr) {
-			log.Tracef("Add %s to whitelist", addr)
-			AddToWl(dc.addr, false)
+		// handle dialing result
+		go func() {
+			t := time.NewTimer(TimeoutToConnect)
+			defer t.Stop()
+			for i := 0; i < 2; i++ {
+				log.Tracef("Waiting for connection to %s, round %d", dc.addr, i)
+				select {
+				case c := <-ch:
+					if c == nil {
+						log.Tracef("No new connection to %s remaining, return", dc.addr)
+						return
+					}
+					// first connection made, pass it back to caller
+					if i == 0 {
+						dc.conns <- c
+						chAnyConn <- true
+					} else {
+						if c.ConnType() == connTypeDirect {
+							// Could happen if direct route is much slower.
+							log.Debugf("Direct connection to %s established too late, close it", dc.addr)
+							_ = c.Close()
+							return
+						}
+						log.Tracef("Feed detour connection to %s to read/write op", dc.addr)
+						dc.chDetourConn <- c
```

Comment:
When a connection is received and it is not direct, we'll write to this channel and then the loop will exit, because `i` is now `1`. It would probably be safer to end this loop explicitly with `return`. My reasoning is that if `i` would be allowed to grow to `3` or more, then this loop will continue one more time at least, the problem is we already have what we wanted.